common/src/main/java/io/netty/util/AsciiString.java

```diff
@@ -36,10 +39,14 @@
  * {@link ByteBuffer}. It is often used in conjunction with {@link TextHeaders}.
  */
 public final class AsciiString extends ByteString implements CharSequence, Comparable<CharSequence> {
-
+    private static final HashCodeGenerator CASE_INSENSITIVE_HASHER =
+            PlatformDependent.hashCodeGeneratorAsciiCaseInsensitive();
     private static final char MAX_CHAR_VALUE = 255;
     public static final AsciiString EMPTY_STRING = new AsciiString("");
 
+    private String string;
+    private int caseInsensitiveHashCode = CASE_INSENSITIVE_HASHER.emptyHashValue();
```

Comment:
SGTM just wanted to discuss.  I think it's fine keeping it non-volatile since we'll usually be accessing from a single thread anyway. If we run into caching issues, recomputing won't kill us (so long as the result is always the same). If the array changes that's a different story but I don't think we have to be terribly concerned about it at this point.